File: ts/src/DeviceLink.ts
import { getPublicKey, generateSecretKey } from "nostr-tools";

import { Invite } from "./Invite";

export interface DeviceLinkRequest {
  requestPubkey: string;
  deviceAppKeyPubkey: string;
  requestSecret: string;
  requestedAt: number;
  deviceLabel?: string;
  clientLabel?: string;
}

export interface LocalDeviceLinkRequest extends DeviceLinkRequest {
  requestSecretKey: Uint8Array;
}

export interface CreatedDeviceLinkRequest {
  request: LocalDeviceLinkRequest;
  deviceAppKeySecretKey: Uint8Array;
  code: string;
}

export function createDeviceLinkRequest(
  options: {
    deviceAppKeySecretKey?: Uint8Array;
    requestSecretKey?: Uint8Array;
    requestedAt?: number;
    deviceLabel?: string;
    clientLabel?: string;
  } = {},
): CreatedDeviceLinkRequest {
  const deviceAppKeySecretKey =
    options.deviceAppKeySecretKey ?? generateSecretKey();
  const requestSecretKey = options.requestSecretKey ?? generateSecretKey();
  const requestSecret = hexFromBytes(requestSecretKey);
  const request: LocalDeviceLinkRequest = {
    requestPubkey: getPublicKey(requestSecretKey),
    requestSecretKey,
    deviceAppKeyPubkey: getPublicKey(deviceAppKeySecretKey),
    requestSecret,
    requestedAt: options.requestedAt ?? currentUnixSeconds(),
    ...(normalizeLabel(options.deviceLabel)
      ? { deviceLabel: normalizeLabel(options.deviceLabel)! }
      : {}),
    ...(normalizeLabel(options.clientLabel)
      ? { clientLabel: normalizeLabel(options.clientLabel)! }
      : {}),
  };
  return {
    request,
    deviceAppKeySecretKey,
    code: encodeCompactDeviceLinkRequest(request),
  };
}

export function encodeCompactDeviceLinkRequest(request: {
  deviceAppKeyPubkey: string;
  requestSecret?: string;
  requestSecretKey?: Uint8Array;
  requestedAt?: number;
  deviceLabel?: string;
  clientLabel?: string;
}): string {
  const requestSecret = request.requestSecretKey
    ? hexFromBytes(request.requestSecretKey)
    : request.requestSecret;
  if (
    !requestSecret ||
    !isHex32Bytes(request.deviceAppKeyPubkey) ||
    !isHex32Bytes(requestSecret)
  ) {
    throw new Error("Invalid compact device link request");
  }
  const metadata = encodeDeviceLinkMetadata({
    requestedAt: request.requestedAt,
    deviceLabel: request.deviceLabel,
    clientLabel: request.clientLabel,
  });
  return `${request.deviceAppKeyPubkey.toLowerCase()}.${requestSecret.toLowerCase()}.${metadata}`;
}

export function parseCompactDeviceLinkRequest(
  input: string,
): DeviceLinkRequest | null {
  const parts = input.trim().split(".");
  if (parts.length !== 3) return null;
  const [deviceAppKeyPubkey, requestSecret, encodedMetadata] = parts;
  if (!isHex32Bytes(deviceAppKeyPubkey) || !isHex32Bytes(requestSecret))
    return null;
  const metadata = decodeDeviceLinkMetadata(encodedMetadata);
  if (!metadata) return null;
  const requestSecretKey = bytesFromHex(requestSecret);
  try {
    getPublicKey(requestSecretKey);
  } catch {
    return null;
  }
  return {
    requestPubkey: getPublicKey(requestSecretKey),
    deviceAppKeyPubkey: deviceAppKeyPubkey.toLowerCase(),
    requestSecret: requestSecret.toLowerCase(),
    requestedAt: metadata.requestedAt ?? currentUnixSeconds(),
    ...(metadata.deviceLabel ? { deviceLabel: metadata.deviceLabel } : {}),
    ...(metadata.clientLabel ? { clientLabel: metadata.clientLabel } : {}),
  };
}

export function deterministicLinkInviteForDeviceLinkRequest(
  request: DeviceLinkRequest,
): Invite {
  if (
    !isHex32Bytes(request.deviceAppKeyPubkey) ||
    !isHex32Bytes(request.requestSecret)
  ) {
    throw new Error("Invalid compact device link request");
  }
  const rng = new StdRngCompat(bytesFromHex(request.requestSecret));
  const inviterEphemeralPrivateKey = rng.nextSecretKey();
  const sharedSecret = hexFromBytes(rng.nextSecretKey());
  const deviceAppKeyPubkey = request.deviceAppKeyPubkey.toLowerCase();
  return new Invite(
    getPublicKey(inviterEphemeralPrivateKey),
    sharedSecret,
    deviceAppKeyPubkey,
    inviterEphemeralPrivateKey,
    deviceAppKeyPubkey,
    1,
    [],
    0,
    "link",
  );
}

const currentUnixSeconds = (): number => Math.floor(Date.now() / 1000);

const isHex32Bytes = (value: string): boolean => /^[0-9a-f]{64}$/i.test(value);

interface CompactDeviceLinkMetadata {
  v: 1;
  requestedAt?: number;
  deviceLabel?: string;
  clientLabel?: string;
}

const encodeDeviceLinkMetadata = (input: {
  requestedAt?: number;
  deviceLabel?: string;
  clientLabel?: string;
}): string => {
  const requestedAt =
    input.requestedAt !== undefined &&
    Number.isInteger(input.requestedAt) &&
    input.requestedAt >= 0
      ? Math.floor(input.requestedAt)
      : undefined;
  const metadata: CompactDeviceLinkMetadata = {
    v: 1,
    ...(requestedAt !== undefined ? { requestedAt } : {}),
    ...(normalizeLabel(input.deviceLabel)
      ? { deviceLabel: normalizeLabel(input.deviceLabel)! }
      : {}),
    ...(normalizeLabel(input.clientLabel)
      ? { clientLabel: normalizeLabel(input.clientLabel)! }
      : {}),
  };
  return base64UrlEncodeUtf8(JSON.stringify(metadata));
};

const decodeDeviceLinkMetadata = (
  input: string | undefined,
): CompactDeviceLinkMetadata | null => {
  if (!input || !/^[A-Za-z0-9_-]+$/.test(input)) return null;
  try {
    const value = JSON.parse(
      base64UrlDecodeUtf8(input),
    ) as Partial<CompactDeviceLinkMetadata>;
    if (!value || value.v !== 1) return null;
    const requestedAt =
      value.requestedAt !== undefined &&
      Number.isInteger(value.requestedAt) &&
      value.requestedAt >= 0
        ? value.requestedAt
        : undefined;
    return {
      v: 1,
      ...(requestedAt !== undefined ? { requestedAt } : {}),
      ...(normalizeLabel(value.deviceLabel)
        ? { deviceLabel: normalizeLabel(value.deviceLabel)! }
        : {}),
      ...(normalizeLabel(value.clientLabel)
        ? { clientLabel: normalizeLabel(value.clientLabel)! }
        : {}),
    };
  } catch {
    return null;
  }
};

const normalizeLabel = (label: unknown): string | undefined => {
  if (typeof label !== "string") return undefined;
  const normalized = label.trim().replace(/\s+/g, " ");
  if (!normalized) return undefined;
  return Array.from(normalized).slice(0, 96).join("");
};

const textEncoder = new TextEncoder();
const textDecoder = new TextDecoder();

const base64UrlEncodeUtf8 = (value: string): string => {
  const bytes = textEncoder.encode(value);
  let binary = "";
  for (const byte of bytes) {
    binary += String.fromCharCode(byte);
  }
  return btoa(binary)
    .replace(/\+/g, "-")
    .replace(/\//g, "_")
    .replace(/=+$/, "");
};

const base64UrlDecodeUtf8 = (value: string): string => {
  const padded = value
    .replace(/-/g, "+")
    .replace(/_/g, "/")
    .padEnd(Math.ceil(value.length / 4) * 4, "=");
  const binary = atob(padded);
  const bytes = new Uint8Array(binary.length);
  for (let index = 0; index < binary.length; index += 1) {
    bytes[index] = binary.charCodeAt(index);
  }
  return textDecoder.decode(bytes);
};

const hexFromBytes = (bytes: Uint8Array): string =>
  Array.from(bytes)
    .map((byte) => byte.toString(16).padStart(2, "0"))
    .join("");

const bytesFromHex = (hex: string): Uint8Array => {
  if (!isHex32Bytes(hex)) throw new Error("Invalid hex secret");
  const bytes = new Uint8Array(32);
  for (let index = 0; index < bytes.length; index += 1) {
    bytes[index] = Number.parseInt(hex.slice(index * 2, index * 2 + 2), 16);
  }
  return bytes;
};

class StdRngCompat {
  private readonly seed: Uint8Array;
  private blockCounter = 0n;
  private buffer = new Uint8Array();
  private offset = 0;

  constructor(seed: Uint8Array) {
    if (seed.length !== 32) throw new Error("StdRng seed must be 32 bytes");
    this.seed = seed;
  }

  nextSecretKey(): Uint8Array {
    for (let attempts = 0; attempts < 16; attempts += 1) {
      const candidate = this.nextBytes(32);
      try {
        getPublicKey(candidate);
        return candidate;
      } catch {
        // Try another deterministic block.
      }
    }
    throw new Error("Could not derive a valid deterministic link key");
  }

  private nextBytes(length: number): Uint8Array {
    const output = new Uint8Array(length);
    let written = 0;
    while (written < length) {
      if (this.offset >= this.buffer.length) {
        this.buffer = this.nextBlock();
        this.offset = 0;
      }
      const take = Math.min(length - written, this.buffer.length - this.offset);
      output.set(this.buffer.slice(this.offset, this.offset + take), written);
      this.offset += take;
      written += take;
    }
    return output;
  }

  private nextBlock(): Uint8Array<ArrayBuffer> {
    const block = chachaBlock(this.seed, this.blockCounter, 12);
    this.blockCounter += 1n;
    return block;
  }
}

const chachaBlock = (
  key: Uint8Array,
  counter: bigint,
  rounds: number,
): Uint8Array<ArrayBuffer> => {
  const state = new Uint32Array(16);
  state[0] = 0x61707865;
  state[1] = 0x3320646e;
  state[2] = 0x79622d32;
  state[3] = 0x6b206574;
  for (let index = 0; index < 8; index += 1) {
    state[4 + index] = readU32Le(key, index * 4);
  }
  state[12] = Number(counter & 0xffffffffn);
  state[13] = Number((counter >> 32n) & 0xffffffffn);
  state[14] = 0;
  state[15] = 0;

  const working = new Uint32Array(state);
  for (let index = 0; index < rounds / 2; index += 1) {
    quarterRound(working, 0, 4, 8, 12);
    quarterRound(working, 1, 5, 9, 13);
    quarterRound(working, 2, 6, 10, 14);
    quarterRound(working, 3, 7, 11, 15);
    quarterRound(working, 0, 5, 10, 15);
    quarterRound(working, 1, 6, 11, 12);
    quarterRound(working, 2, 7, 8, 13);
    quarterRound(working, 3, 4, 9, 14);
  }

  const output = new Uint8Array(64);
  for (let index = 0; index < 16; index += 1) {
    writeU32Le(output, index * 4, (working[index] + state[index]) >>> 0);
  }
  return output;
};

const quarterRound = (
  state: Uint32Array,
  a: number,
  b: number,
  c: number,
  d: number,
): void => {
  state[a] = (state[a] + state[b]) >>> 0;
  state[d] = rotateLeft(state[d] ^ state[a], 16);
  state[c] = (state[c] + state[d]) >>> 0;
  state[b] = rotateLeft(state[b] ^ state[c], 12);
  state[a] = (state[a] + state[b]) >>> 0;
  state[d] = rotateLeft(state[d] ^ state[a], 8);
  state[c] = (state[c] + state[d]) >>> 0;
  state[b] = rotateLeft(state[b] ^ state[c], 7);
};

const rotateLeft = (value: number, shift: number): number =>
  ((value << shift) | (value >>> (32 - shift))) >>> 0;

const readU32Le = (bytes: Uint8Array, offset: number): number =>
  (bytes[offset] |
    (bytes[offset + 1] << 8) |
    (bytes[offset + 2] << 16) |
    (bytes[offset + 3] << 24)) >>>
  0;

const writeU32Le = (bytes: Uint8Array, offset: number, value: number): void => {
  bytes[offset] = value & 0xff;
  bytes[offset + 1] = (value >>> 8) & 0xff;
  bytes[offset + 2] = (value >>> 16) & 0xff;
  bytes[offset + 3] = (value >>> 24) & 0xff;
};

File: rust/crates/nostr-double-ratchet/src/device_link.rs
use crate::{DevicePubkey, Error, Invite, ProtocolContext, Result, UnixSeconds};
use base64::{engine::general_purpose::URL_SAFE_NO_PAD, Engine};
use nostr::{Keys, PublicKey, SecretKey};
use rand::rngs::StdRng;
use rand::SeedableRng;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DeviceLinkRequest {
    pub request_pubkey: PublicKey,
    pub device_app_key_pubkey: PublicKey,
    pub request_secret: String,
    pub requested_at: Option<u64>,
    pub device_label: Option<String>,
    pub client_label: Option<String>,
}

pub fn encode_compact_device_link_request(
    device_app_key_pubkey: PublicKey,
    request_secret: &str,
    device_label: Option<&str>,
    client_label: Option<&str>,
    requested_at: Option<u64>,
) -> Result<String> {
    let secret = parse_secret_hex(request_secret)?;
    let metadata = CompactDeviceLinkMetadata {
        v: 1,
        requested_at,
        device_label: normalize_label(device_label),
        client_label: normalize_label(client_label),
    };
    let metadata = URL_SAFE_NO_PAD.encode(serde_json::to_vec(&metadata)?);
    Ok(format!(
        "{}.{}.{}",
        device_app_key_pubkey.to_hex(),
        hex::encode(secret),
        metadata
    ))
}

pub fn parse_compact_device_link_request(input: &str) -> Result<DeviceLinkRequest> {
    let mut parts = input.trim().split('.');
    let Some(device_app_key_pubkey) = parts.next() else {
        return Err(Error::Parse("Missing device app key pubkey".to_string()));
    };
    let Some(request_secret) = parts.next() else {
        return Err(Error::Parse("Missing device link secret".to_string()));
    };
    let Some(metadata) = parts.next() else {
        return Err(Error::Parse("Missing device link metadata".to_string()));
    };
    if parts.next().is_some() {
        return Err(Error::Parse(
            "Invalid compact device link request".to_string(),
        ));
    }

    let device_app_key_pubkey = PublicKey::parse(device_app_key_pubkey)
        .map_err(|error| Error::Parse(format!("Invalid device app key pubkey: {error}")))?;
    let request_secret_bytes = parse_secret_hex(request_secret)?;
    let metadata = parse_device_link_metadata(metadata)?;
    let request_secret_key = SecretKey::from_slice(&request_secret_bytes)?;
    let request_pubkey = Keys::new(request_secret_key).public_key();

    Ok(DeviceLinkRequest {
        request_pubkey,
        device_app_key_pubkey,
        request_secret: hex::encode(request_secret_bytes),
        requested_at: metadata.requested_at,
        device_label: metadata.device_label,
        client_label: metadata.client_label,
    })
}

pub fn deterministic_link_invite_for_device_link_request(
    request: &DeviceLinkRequest,
) -> Result<Invite> {
    deterministic_link_invite_for_device(
        request.device_app_key_pubkey,
        request.request_secret.as_str(),
    )
}

pub fn deterministic_link_invite_for_device(
    device_app_key_pubkey: PublicKey,
    request_secret: &str,
) -> Result<Invite> {
    let seed = parse_secret_hex(request_secret)?;
    let mut rng = StdRng::from_seed(seed);
    let mut ctx = ProtocolContext::new(UnixSeconds(0), &mut rng);
    let mut invite = Invite::create_new_with_context(
        &mut ctx,
        DevicePubkey::from_bytes(device_app_key_pubkey.to_bytes()),
        None,
        Some(1),
    )?;
    invite.purpose = Some("link".to_string());
    invite.device_id = Some(device_app_key_pubkey.to_hex());
    Ok(invite)
}

fn parse_secret_hex(secret: &str) -> Result<[u8; 32]> {
    let secret = secret.trim().to_ascii_lowercase();
    if secret.len() != 64 {
        return Err(Error::Parse(
            "Device link secret must be 32 bytes".to_string(),
        ));
    }
    let bytes = hex::decode(secret)?;
    let bytes: [u8; 32] = bytes
        .try_into()
        .map_err(|_| Error::Parse("Device link secret must be 32 bytes".to_string()))?;
    SecretKey::from_slice(&bytes)?;
    Ok(bytes)
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct CompactDeviceLinkMetadata {
    v: u8,
    #[serde(
        rename = "requestedAt",
        default,
        skip_serializing_if = "Option::is_none"
    )]
    requested_at: Option<u64>,
    #[serde(
        rename = "deviceLabel",
        default,
        skip_serializing_if = "Option::is_none"
    )]
    device_label: Option<String>,
    #[serde(
        rename = "clientLabel",
        default,
        skip_serializing_if = "Option::is_none"
    )]
    client_label: Option<String>,
}

fn parse_device_link_metadata(input: &str) -> Result<CompactDeviceLinkMetadata> {
    if input.is_empty()
        || !input
            .bytes()
            .all(|byte| byte.is_ascii_alphanumeric() || byte == b'-' || byte == b'_')
    {
        return Err(Error::Parse("Invalid device link metadata".to_string()));
    }
    let bytes = URL_SAFE_NO_PAD
        .decode(input)
        .map_err(|error| Error::Parse(format!("Invalid device link metadata: {error}")))?;
    let metadata: CompactDeviceLinkMetadata = serde_json::from_slice(&bytes)?;
    if metadata.v != 1 {
        return Err(Error::Parse("Unsupported device link metadata".to_string()));
    }
    Ok(CompactDeviceLinkMetadata {
        v: 1,
        requested_at: metadata.requested_at,
        device_label: normalize_label(metadata.device_label.as_deref()),
        client_label: normalize_label(metadata.client_label.as_deref()),
    })
}

fn normalize_label(label: Option<&str>) -> Option<String> {
    let normalized = label?.split_whitespace().collect::<Vec<_>>().join(" ");
    if normalized.is_empty() {
        return None;
    }
    Some(normalized.chars().take(96).collect())
}
